### PR TITLE
vfp for ecl2df

### PR DESCRIPTION
**Pull request by runhol**
Extension of ecl2df for vfp curves

### CODE REVIEW

File: ecl2df/vfp.py [hunk 1]
```diff
@@ -0,0 +1,1413 @@
...
+
+    # This import is seemingly not used, but necessary for some attributes
+    # to be included in DeckItem objects.
+    from opm.io.deck import DeckKeyword  # noqa
+except ImportError:
+    pass
+
+from ecl2df import EclFiles, common, getLogger_ecl2csv
+
+logger = logging.getLogger(__name__)
```
[the remark this answers is not in the document]
Comment (lindjoha) [reply ecl2df/vfp.py:31]: @runhol

[the remark this answers is not in the document]
Comment (berland) [reply ecl2df/vfp.py:31]: `logger` is listed as an exception for `.pylintrc`. You can change it, but then it should be done in all other modules as well.

File: ecl2df/vfp.py [hunk 2]
```diff
@@ -0,0 +1,1413 @@
...
+]
+
+# The renamers listed here map from opm-common json item names to
+# desired column names in produced dataframes. They also to a certain
+# extent determine the structure of the dataframe, in particular
+# for keywords with arbitrary data amount pr. record (GAS, THP, WGR, GOR f.ex)
+RENAMERS: Dict[str, Dict[str, Union[str, List[str]]]] = {}
+
+
+# Type of VFP curve
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py:45]: Enum classes and unit type definitions kept within vfp.py for now to be consistent with other ecl2df modules. Can be extracted to a separate file later.

File: ecl2df/vfp.py [hunk 3]
```diff
@@ -0,0 +1,1413 @@
...
+    flow_values: Union[Any, List[float]]
+    thp_values: Union[Any, List[float]]
+    wfr_values: Union[Any, List[float]]
+    gfr_values: Union[Any, List[float]]
+    alq_values: Union[Any, List[float]]
+    # Extract interpolation ranges into lists
+    if isinstance(flow_record.get("FLOW_VALUES"), list):
+        flow_values = flow_record.get("FLOW_VALUES")
+    elif isinstance(flow_record.get("FLOW_VALUES"), numbers.Number):
+        flow_values = [flow_record.get("FLOW_VALUES")]
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Function deckrecord2list created to extract list from a record.

File: ecl2df/vfp.py [hunk 4]
```diff
@@ -0,0 +1,1413 @@
...
+        thp_value = thp_values[thp_index]
+
+        bhp_record_values = [thp_value] + bhp_values
+        if len(bhp_values) != no_flow_values:
+            logger.error(
+                "Dimension of record of tabulated values "
+                "does not match number of flow values"
+            )
+            return pd.DataFrame()
+        bhp_array_values.append(bhp_record_values)
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Function vfptable2df implemented to perform stacking of  VFP table data into DataFrame. Used in vfpprod2df and vfpinj2df.

File: ecl2df/vfp.py [hunk 5]
```diff
@@ -0,0 +1,1413 @@
...
+    Args:
+        table: DataFrame with multiindex for table ranges and colums
+               for tabulated values (BHP)
+        format: Format string for values
+        values_per_line: Number of values per line in output
+    """
+
+    ecl_str = ""
+    for idx, row in table.iterrows():
+        ecl_str += "%2d" % (idx)
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Use of f string implemented.

File: ecl2df/vfp.py [hunk 6]
```diff
@@ -0,0 +1,1413 @@
...
+    Args:
+        table: DataFrame with multiindex for table ranges and colums
+               for tabulated values (BHP)
+        format: Format string for values
+        values_per_line: Number of values per line in output
+    """
+
+    ecl_str = ""
+    for idx, row in table.iterrows():
+        ecl_str += "%2d %2d %2d %2d" % (idx[0], idx[1], idx[2], idx[3])
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Use of f string implemented.

File: ecl2df/vfp.py [hunk 7]
```diff
@@ -0,0 +1,1413 @@
...
+#        logger.warning('Empty dataframe, input file does not contain keywords')
+
+
+def fill_parser(parser: argparse.ArgumentParser) -> argparse.ArgumentParser:
+    """Set up sys.argv parsers.
+
+    Arguments:
+        parser (argparse.ArgumentParser or argparse.subparser): parser
+            to fill with arguments
+    """
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py:1381]: List of vfp table numbers to export implemented. Input given as string, example "[0,1,3:7]" (=[0,1,3,4,5,6,7]). Also input of VFP type changes from list to only accept either 'VFPPROD' or 'VFPINJ'.

File: tests/test_vfp.py [hunk 8]
```diff
@@ -0,0 +1,71 @@
...
+    )
+
+TESTDIR = Path(__file__).absolute().parent
+VFPDIR = str(TESTDIR) + "/data/vfp/"
+VFPPRODFILES = ["/test_vfpprod.Ecl", "test_vfpprod_defaulted.Ecl"]
+VFPINJFILES = ["test_vfpinj.Ecl", "test_vfpinj_defaulted.Ecl"]
+CSVPRODFILES = ["test_vfpprod.csv", "test_vfpprod_defaulted.csv"]
+CSVINJFILES = ["test_vfpinj.csv", "test_vfpinj_defaulted.csv"]
+
+
```
[the remark this answers is not in the document]
Comment (runhol) [reply tests/test_vfp.py]: Tests implemented in test_vfp.py using @pytest.mark.parameterize. Additional tests implemented. Eclipse VFP files and csv files under data/vfp used for previous tests removed.

File: ecl2df/vfp.py [hunk 9]
```diff
@@ -0,0 +1,1413 @@
...
+        return []
+
+    vfp_keywords = []
+    if keywords:
+        for keyword in keywords:
+            if keyword not in SUPPORTED_KEYWORDS:
+                logger.warning(
+                    "Given keyword %s is not in supported keywords" % (keyword)
+                )
+            else:
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: keyword defaulted to "VFPPROD". Only "VFPPROD" or "VFPINJ" allowed as input.

File: ecl2df/vfp.py [hunk 10]
```diff
@@ -0,0 +1,1413 @@
...
+    no_flow_values = len(flow_values)
+    no_thp_values = len(thp_values)
+    no_interp_values = no_thp_values
+    no_tab_records = num_rec - 3
+    if no_interp_values != no_tab_records:
+        logger.error(
+            "Dimensions of interpolation ranges "
+            "does not match number of tabulated records"
+        )
+        return pd.DataFrame()
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: logger.error replaced by raise ValueError

File: ecl2df/vfp.py [hunk 11]
```diff
@@ -0,0 +1,1413 @@
...
+            "UNIT_TYPE",
+        ]
+    ]
+
+    return df_bhp_stacked
+
+
+def dfs(
+    deck: Union[str, EclFiles, "opm.libopmcommon_python.Deck"],
+    keywords: Optional[List[str]] = ["VFPROD"],
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Argument replaced by string that should be either "VFPPROD" or "VFPINJ".

File: ecl2df/vfp.py [hunk 12]
```diff
@@ -0,0 +1,1413 @@
...
+    df_bhp_stacked.bfill(inplace=True)
+    # Also reset the index:
+    df_bhp_stacked.reset_index(inplace=True)
+    df_bhp_stacked.drop("level_0", axis="columns", inplace=True)
+    # This column is not meaningful (it is the old index)
+
+    # Delete rows that does not belong to any flow rate (this is
+    # possibly a by-product of not doing the stacking in an
+    # optimal way)
+    df_bhp_stacked = df_bhp_stacked[df_bhp_stacked["level_1"] != "DELETE"]
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Stacking of VFP table data into DataFrame put in function vfptable2df. Function used by vfpprod2df and vfpinj2df. Separate testing to be included.

File: ecl2df/vfp.py [hunk 13]
```diff
@@ -0,0 +1,1413 @@
...
+            "does not match number of tabulated records"
+        )
+        return pd.DataFrame()
+
+    # Extract basic table information
+    table = int(basic_record["TABLE"])
+    datum = float(basic_record["DATUM_DEPTH"])
+    rate = VFPPROD_FLO.GAS
+    if basic_record["RATE_TYPE"]:
+        rate = VFPPROD_FLO[basic_record["RATE_TYPE"]]
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py:475]: With defaulted rate the return value is None which should be mapped to VFPPROD_FLO.GAS. Cannot map both None and "GAS" to VFPPROD_FLO.GAS in class VFPPROD_FLO(Enum). The above test sets the correct rate type if this parameter is defaulted (=None).

File: ecl2df/vfp.py [hunk 14]
```diff
@@ -0,0 +1,1413 @@
...
+class ALQ(Enum):
+    GRAT = "GRAT"
+    IGLR = "IGLR"
+    TGLR = "TGLR"
+    PUMP = "PUMP"
+    COMP = "COMP"
+    DENO = "DENO"
+    DENG = "DENG"
+    BEAN = "BEAN"
+    UNDEFINED = "UNDEFINED"
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py]: Maybe keep UNDEFINED here, but add additional tests when writing to csv to output empty column when ALQ is undefined?

File: ecl2df/vfp.py [hunk 15]
```diff
@@ -0,0 +1,1446 @@
...
+    dfs_vfp = []
+    # The keywords VFPPROD/VFPINJ can be used many times in Eclipse and be introduced in
+    # separate files or a common file. Need to loop to find all instances of keyword and
+    # store separately
+    for deck_keyword in deck:
+        if deck_keyword.name == keyword:
+            if deck_keyword.name == "VFPPROD":
+                df_vfpprod = vfpprod2df(deck_keyword)
+                if not vfpnumbers or df_vfpprod.loc[0, "TABLE_NUMBER"] in vfpnumbers:
+                    dfs_vfp.append(df_vfpprod)
```
Comment (lindjoha) [inline ecl2df/vfp.py:741]: Is there any way of finding the table number without reading the whole table?

File: ecl2df/vfp.py [hunk 16]
```diff
@@ -0,0 +1,1446 @@
...
+            "GAS": "sm3/day",
+            "WG": "sm3/day",
+            "TM": "kg-M/day",
+        },
+        "THP": {"THP": "atma"},
+    },
+}
+
+
+def deckrecord2list(
```
Comment (lindjoha) [inline ecl2df/vfp.py:343]: underscore notation ? `_deckrecord2list()`

File: ecl2df/vfp.py [hunk 17]
```diff
@@ -0,0 +1,1446 @@
...
+            "ALQ_TYPE",
+            "PRESSURE_TYPE",
+            "TAB_TYPE",
+            "UNIT_TYPE",
+        ]
+    ]
+
+    # reset index (not used other than tests)
+    df_bhp_stacked.reset_index(inplace=True, drop=True)
+    return df_bhp_stacked
```
Comment (lindjoha) [inline ecl2df/vfp.py:595]: This is details, but maybe more elegant to write:

`return df_bhp_stacked.reset_index(drop=True)`

**Review by lindjoha — CHANGES_REQUESTED** [review]
Looks very good now! Only some minor comments.

File: ecl2df/csv2ecl.py [hunk 18]
```diff
@@ -89,6 +89,16 @@ def get_parser() -> argparse.ArgumentParser:
     satfunc.fill_reverse_parser(satfunc_parser)
     satfunc_parser.set_defaults(func=satfunc.satfunc_reverse_main)
 
+    vfp_parser = subparsers.add_parser(
+        "vfp",
+        help="Write VFPPROD/VFPINJ include files",
+        description=(
+            "Write VFPPROD/VFPINJ include files from CSV files on " "the ecl2df format."
+        ),
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/csv2ecl.py]: Corrected and splitted on two lines.

**Comment by lindjoha** [comment]
I suggest that you implement an option to export the data on `pyarrow` format. See `summary` or `wellcompletiondata` module for examples.

**Comment by lindjoha** [comment]
I discussed with @asnyv the possibility of exporting the VFPs as separate files. Or possibly as "subtables" in one arrow file if that is possible. There is many advantages to this. Types and units could be put as column metadata instead of set per row, this would reduce the amount of data significantly. In webviz you could hold less data in memory and read tables on demand.

**Comment by asnyv** [comment]
Another idea is to maybe move all the metadata into the metadata of an arrow column (including the table number, maybe the filename, well/pipeline variables, like rate, THP, GOR, WGR and ALQ values), and then stack the output (called the "BHP"s even though that is only valid for wells). Then you can have one arrow file as a "database" of all VFPs, and then fetch single columns containing all the data for one VFP table to minimize duplicated data and limit data transfer (at least from webviz point of view, for a manual user in e.g. excel, the current csv format is probably easier to handle).

Think we should discuss this with @sigurdp after the summer, as I'm sure he has some valuable input on optimizing this structure.

File: ecl2df/vfp.py [hunk 19]
```diff
@@ -0,0 +1,2651 @@
...
+    col_metadata_list = []
+    num_records = len(thp_values) * len(wfr_values) * len(gfr_values) * len(alq_values)
+    for i in range(0, num_records):
+        thp_idx = thp_indices[i]
+        wfr_idx = wfr_indices[i]
+        gfr_idx = gfr_indices[i]
+        alq_idx = alq_indices[i]
+        col_name = (
+            str(thp_idx) + "_" + str(wfr_idx) + "_" + str(gfr_idx) + "_" + str(alq_idx)
+        )
```
Comment (lindjoha) [inline ecl2df/vfp.py:746]: Use f-string

File: ecl2df/vfp.py [hunk 20]
```diff
@@ -0,0 +1,2651 @@
...
+        tab_data    : tabulated (BHP) data
+                      (ordered as thp-, wfr-, gfr-, alq- and flow-values)
+    """
+
+    # Generate list with values instead of indices
+    thp_values_list = [thp_values[i - 1] for i in thp_indices]
+    wfr_values_list = [wfr_values[i - 1] for i in wfr_indices]
+    gfr_values_list = [gfr_values[i - 1] for i in gfr_indices]
+    alq_values_list = [alq_values[i - 1] for i in alq_indices]
+
```
Comment (lindjoha) [inline ecl2df/vfp.py:544]: Is the point of this that you need lists starting from index 1 ?

File: ecl2df/vfp.py [hunk 21]
```diff
@@ -0,0 +1,2651 @@
...
+
+def check_vfpprod_basic_data(vfp_data: Dict[str, Any]) -> bool:
+    """Perform a check of the VFPPROD data contained in the dictionary.
+    Checks if all data is present and if the dimensions of the arrays
+    are consisitent.
+
+    Args:
+        vfp_data:   Dictionary containing all data for a VFPPROD keyword in Eclipse
+    """
```
Comment (lindjoha) [inline ecl2df/vfp.py:1425]: Make function private?

File: ecl2df/vfp.py [hunk 22]
```diff
@@ -0,0 +1,2651 @@
...
+    no_records = no_thp_values * no_wfr_values * no_gfr_values * no_alq_values
+    if no_tab_values % no_flow_values > 0:
+        return False
+    elif no_tab_values // no_flow_values != no_records:
+        return False
+
+    return True
+
+
+def check_vfpinj_basic_data(vfp_data: Dict[str, Any]) -> bool:
```
Comment (lindjoha) [inline ecl2df/vfp.py:1494]: Make private?

File: ecl2df/vfp.py [hunk 23]
```diff
@@ -0,0 +1,2651 @@
...
+                pa_table_vfpinj = vfpinj2pyarrow(deck_keyword, vfpnumbers_str)
+                if pa_table_vfpinj is not None:
+                    pyarrow_tables_vfp.append(pa_table_vfpinj)
+
+    return pyarrow_tables_vfp
+
+
+def write_eclipse_comment(comment: str, max_char_per_line: int = 72) -> str:
+    """Produce a string representing comment in Eclipse file
+
```
Comment (lindjoha) [inline ecl2df/vfp.py:1794]: make private?

File: ecl2df/vfp.py [hunk 24]
```diff
@@ -0,0 +1,2651 @@
...
+                else:
+                    new_line += words[0]
+                words.pop(0)
+            comment_str += new_line + "\n"
+
+    return comment_str
+
+
+def write_vfpprod_basic_record(
+    tableno: int,
```
Comment (lindjoha) [inline ecl2df/vfp.py:1822]: Make private?

File: ecl2df/vfp.py [hunk 25]
```diff
@@ -0,0 +1,2651 @@
...
+    ecl_str += f"  {wfr_type:>8s}"
+    ecl_str += f"  {gfr_type:>8s}"
+    ecl_str += f"  {pressure_type:>8s}"
+    ecl_str += f"  {alq_type_str:>8s}"
+    ecl_str += f"  {unit_type:>6s}"
+    ecl_str += f"  {tab_type:>8s} /\n\n"
+    return ecl_str
+
+
+def write_vfpinj_basic_record(
```
Comment (lindjoha) [inline ecl2df/vfp.py:1866]: Make private?

File: ecl2df/vfp.py [hunk 26]
```diff
@@ -0,0 +1,2651 @@
...
+    ecl_str += f"   {tableno:5d}"
+    ecl_str += f"  {datum:11.1f}"
+    ecl_str += f"  {flo_type:>9s}"
+    ecl_str += f"  {pressure_type:>8s}"
+    ecl_str += f"  {unit_type_str:>8s}"
+    ecl_str += f"  {tab_type:>8s} /\n\n"
+    return ecl_str
+
+
+def write_vfp_range(
```
Comment (lindjoha) [inline ecl2df/vfp.py:1899]: Make private?

File: ecl2df/vfp.py [hunk 27]
```diff
@@ -0,0 +1,2651 @@
...
+                    ecl_str += "\n"
+            elif n == no_flo - 1:
+                ecl_str += "\n"
+        ecl_str += "/\n"
+
+    return ecl_str
+
+
+def write_vfpprod_table_records(
+    thp_indices: np.ndarray,
```
Comment (lindjoha) [inline ecl2df/vfp.py:1965]: Make private?

**Review by lindjoha — CHANGES_REQUESTED** [review]

File: ecl2df/vfp.py [hunk 28]
```diff
@@ -0,0 +1,2651 @@
...
+                    dfs_vfp.append(df_vfpprod)
+            elif deck_keyword.name == "VFPINJ":
+                df_vfpinj = vfpinj2df(deck_keyword, vfpnumbers_str)
+                if df_vfpinj is not None:
+                    dfs_vfp.append(df_vfpinj)
+
+    return dfs_vfp
+
+
+def pas(
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp.py:1749]: Changed pas to pyarrow_tables.

File: ecl2df/vfp.py (hunk 20 above)
Comment (runhol) [reply ecl2df/vfp.py:544]: For ecl2df this is done to replace index values (for example 1 1 1 1 at the beginning of a BHP value set) by the actual value of THP, WFR, GFR and ALQ (i.e. the value of the first element in these arrays). Do not want to store both index and value when stacking into a DataFrame.

File: ecl2df/vfp/_vfpcommon.py [hunk 29]
```diff
@@ -0,0 +1,203 @@
...
+    pass
+
+from ecl2df import common
+
+logger = logging.getLogger(__name__)
+
+
+def _string2intlist(list_def_str: str) -> List[int]:
+    """Produce a list of int from input string
+
```
[the remark this answers is not in the document]
Comment (runhol) [reply ecl2df/vfp/_vfpcommon.py:32]: It is already tested by giving different values for the VFP numbers to extract in df.

**Review by lindjoha — APPROVED** [review]
LGTM!
